### PR TITLE
Add computed role section and mappings

### LINKED ISSUE #166: Which ARIA synonym role should computedrole return?
Which ARIA synonym role should `computedrole` return?

WebKit and Chromium differ on this new synonym role test:
https://github.com/web-platform-tests/wpt/pull/38941/files

I think, and the Draft PR test asserts, that the primary role names should be `none` and `image`... I think the secondary synonyms (`presentation` and `img`) should be supported as input, but not returned by `computedrole`.

(Details results linked from table header.)
| role/synonym | [WebKit](https://wpt.fyi/results/wai-aria/role/synonym-roles.html?diff&filter=ADC&run_id=5179462059032576&run_id=5171591195918336) | [Chromium](https://wpt.fyi/results/wai-aria/role/synonym-roles.html?diff&filter=ADC&run_id=6008525770194944&run_id=5181218935537664) |
| ------------- | ------------- | ------------- |
| `none`/`presentation` | ✅`none` (Was: `presentation` until ([bug 224617](https://webkit.org/b/224617))  | ✅`none`  |
| `image`/`img` | ✅`image`  | ❌`generic` / ❓`img` ([bug 1428763](https://bugs.chromium.org/p/chromium/issues/detail?id=1428763)) |

If the WG confirms this assertion, the WPT test will result in 1 bug each for WebKit and Chromium. Maybe 2 for Chrome if it's not already tracking the missing `image` role.

We'd also need something in the ARIA spec… pointing the synonym at its primary, or a table defining the `computedrole` for each role in the spec. 

See also: 
- https://github.com/w3c/aria/issues/1887

### PR DESCRIPTION
**Pull request by spectranaut**
Closes #166

The githack is easier to review, because the mapping tables code works (unlike in the PR preview, below): https://raw.githack.com/w3c/core-aam/computed-roles/index.html#mapping_role

# Implementation

* WPT tests: [tracking issue for all role tests](https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/8)
   - https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/10
  - https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/11
  - https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/12
  - https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/14
  - https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/15
  - https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/16
  - https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/17
  - https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/19
  - https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/20
* Implementations: n/a not normative
* Related WebDriver issue: https://github.com/w3c/webdriver/issues/1727

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/core-aam/pull/167.html" title="Last updated on May 15, 2023, 6:18 PM UTC (0651f71)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/core-aam/167/5ea3335...0651f71.html" title="Last updated on May 15, 2023, 6:18 PM UTC (0651f71)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -546,6 +581,9 @@ <h3>Role Mapping Table</h3>
 				</tr>
 				<tr id="role-map-cell">
 					<th><a class="role-reference" href="#cell"><code>cell</code></a></th>
+					<td class="role-computed">
+						<p><code>cell</code></p>
+					</td>
```
Comment (spectranaut) [inline index.html:586]: For these, should we specify something more like:

Use the computed role <p><code>cell</code></p> when the element is in the appropriate context, and the native host language role otherwise.

If we do this, do we need to define a term "context" maybe.

**Comment by spectranaut** [comment]
hey @cookiecrook and @scottaohara, I'd really like your opinion on this layout!

And @cookiecrook, I'd really like to hear your take or expectations for the roles with are context dependent, and the other roles which are affected by the existence of attributes and accessible names. Have you gotten to thinking about those yet?

File: index.html [hunk 2]
```diff
@@ -342,6 +342,10 @@ <h2>General rules</h2>
 			<li><abbr title="User Interface Automation">UIA</abbr>: expose as <code>AriaRole</code> property. The <code>AriaRole property</code> can also support secondary roles using a space as a separator.</li>
 			<li><abbr title="Accessibility Toolkit">ATK</abbr>/<abbr title="Assistive Technology - Service Provider Interface">AT-SPI</abbr>: expose as an object attribute pair (<code>xml-roles:&quot;string&quot;</code>)</li>
 		  </ul>
+    </section>
+		<section id="roleMappingComputedRole">
+		  <h2>Computed Role</h2>
+                  The computed role of an element is a string that represents the role of the element for the purposes of specification comformance testing. User agents MUST provide this role string, for example, in response to the WebDriver function <a href="https://w3c.github.io/webdriver/#get-computed-role">`getComputedRole`</a>. In the role mapping table below, the computed role of the element will be provided for elements for which the computed role string is different than the role itsself.
```
Comment (pkra) [inline index.html:348]: typo "itsself"

```suggestion
                  The computed role of an element is a string that represents the role of the element for the purposes of specification comformance testing. User agents MUST provide this role string, for example, in response to the WebDriver function <a href="https://w3c.github.io/webdriver/#get-computed-role">`getComputedRole`</a>. In the role mapping table below, the computed role of the element will be provided for elements for which the computed role string is different than the role itself.
```

**Comment by cookiecrook** [comment]
We should discuss further how much we need to spec granularly vs leave to WPT edge case tests. 

For example, it would be tedious to spec all the built-in heuristics of what constitutes a valid data `table` context in the engines such as WebKit’s `isDataTable()`:

https://github.com/WebKit/WebKit/blob/main/Source/WebCore/accessibility/AccessibilityTable.cpp#L111

and I expect there will be some discussion of the purposeful implementation differences related to layout list-itis and an expectation of computedrole:

thread: https://twitter.com/cookiecrook/status/1337226933822603270

**Comment by zcorpan** [comment]
> We should discuss further how much we need to spec granularly vs leave to WPT edge case tests.
> 
> For example, it would be tedious to spec all the built-in heuristics of what constitutes a valid data `table` context in the engines such as WebKit’s `isDataTable()`:
> 
> https://github.com/WebKit/WebKit/blob/main/Source/WebCore/accessibility/AccessibilityTable.cpp#L111

I think this would be good to define, and HTML has a *suggestion* for the heuristic: https://html.spec.whatwg.org/multipage/tables.html#the-table-element

> and I expect there will be some discussion of the purposeful implementation differences related to layout list-itis and an expectation of computedrole:
> 
> thread: https://twitter.com/cookiecrook/status/1337226933822603270

Getting interop on this also seems good imo.

File: index.html [hunk 3]
```diff
@@ -989,6 +1093,9 @@ <h3>Role Mapping Table</h3>
 				</tr>
 				<tr id="role-map-img">
 					<th><a class="role-reference" href="#img"><code>img</code></a></th>
+					<td class="role-computed">
+						<p><code>image</code></p>
+					</td>
```
Comment (cookiecrook) [inline index.html:1098]: Patch also needs an whole TR entry for `image`... I guess we hadn't noticed it was missing... (`none` and `presentation` are already included)

Ideally the `img` row would just point to the entire `image` row for all columns, to avoid duplication. Ditto for the `presentation` row pointing to the `none` row.

File: index.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (spectranaut) [reply index.html]: thanks good point :)

File: index.html (hunk 3 above)
Comment (spectranaut) [reply index.html:1098]: That is true. Unfortunately the table, and it's automatic conversion to the summary/details, doesn't make that easy... but to be honest, I can't say I ever use the "single table view" and it's a bit of a pain. I think I'll add an issue to discuss removing it.

**Comment by cookiecrook** [comment]
- https://github.com/w3c/html-aam/issues/464
- https://github.com/w3c/graphics-aam/issues/7
- https://github.com/w3c/svg-aam/issues/24
- https://github.com/w3c/dpub-aam/issues/29

Did I miss any?

File: index.html (hunk 3 above)
Comment (spectranaut) [reply index.html:1099]: I'm going to fix the missing entry for image after I remove the table view, because I'm motivated to remove the table view right now: https://github.com/w3c/core-aam/issues/169

File: .github/pull_request_template.md [hunk 4]
```diff
@@ -6,8 +6,8 @@ Describe Change Here!
...
 
-* WPT tests: [LINK]()
+* WPT tests:
 * Implementations (link to issue or when done, link to commit):
-   * WebKit: [ISSUE]()
-   * Gecko: [ISSUE]()
-   * Blink: [ISSUE]()
+   * WebKit:
+   * Gecko:
+   * Blink:
```
Comment (cookiecrook) [inline .github/pull_request_template.md:13]: I'm not sure implementation bug links are needed. I've been filing bugs on WebKit and Blink ([example](https://bugs.chromium.org/p/chromium/issues/detail?id=1428763)) as I find them with the related WPT test, and Gecko's WebDriver patch is in source but not yet deployed to WPT.

File: .github/pull_request_template.md [hunk 5]
```diff
@@ -6,8 +6,8 @@ Describe Change Here!
 
 # Implementation
 
-* WPT tests: [LINK]()
+* WPT tests:
```
Comment (cookiecrook) [inline .github/pull_request_template.md:9]: - https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/10
- https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/11
- https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/12
- https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/14
- https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/15
- https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/16
- https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/17

And maybe:
- https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/19
- https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/20

File: index.html (hunk 1 above)
Comment (cookiecrook) [reply index.html:587]: Perhaps that could be in the computed role section (referencing required context), rather than in the table for each.

**Review by cookiecrook — APPROVED** [review]

File: index.html [hunk 6]
```diff
@@ -342,6 +342,11 @@ <h2>General rules</h2>
 			<li><abbr title="User Interface Automation">UIA</abbr>: expose as <code>AriaRole</code> property. The <code>AriaRole property</code> can also support secondary roles using a space as a separator.</li>
 			<li><abbr title="Accessibility Toolkit">ATK</abbr>/<abbr title="Assistive Technology - Service Provider Interface">AT-SPI</abbr>: expose as an object attribute pair (<code>xml-roles:&quot;string&quot;</code>)</li>
 		  </ul>
+    </section>
+		<section id="roleMappingComputedRole">
+		  <h2>Computed Role</h2>
+                  <p>The computed role of an element is a string that represents the role of the element for the purposes of specification comformance testing. When an element has a role but is not contained in the required context (for example, an role `listitem` outside of a `list`), the computed role of the element is unspecified.</p>
```
Comment (scottaohara) [inline index.html:348]: maybe even say "used on an element lacking an accessibility parent with a role of `list`".

```suggestion
                  <p>The computed role of an element is a string that represents the role of the element for the purposes of specification comformance testing. When an element has a role but is not contained in the required context (for example, a role `listitem` used on an element outside of a `list`), the computed role of the element is unspecified.</p>
```

[the remark this answers is not in the document]
Comment (cookiecrook) [reply index.html:348]: I recall writing some tests for orphaned elements, like `<div role="listitem">` but I agree that we do not fully specify or test this yet. 

For example, in the scenario:

```html
<div role="listitemcheckbox checkbox"></div>
```

It's not `listitemcheckbox` b/c of the missing context, but do we use the fallback concrete role, `checkbox`, or the native host language role, `generic`?

In any case, I agree its okay to leave unspecified in this PR.

**Review by cookiecrook — APPROVED** [review]
